domain/secret/watcher_test.go

```diff
@@ -249,20 +233,155 @@ func (s *watcherSuite) TestWatchObsoleteForUnitsOwned(c *gc.C) {
 	wCSingleUnit.AssertChange([]string(nil)...)
 
 	// create revision 2, and obsolete revision 1.
-	createNewRevision(c, uri1)
-	createNewRevision(c, uri2)
+	createNewRevision(c, st, uri1)
+	createNewRevision(c, st, uri2)
 
 	wCSingleUnit.AssertChange(
 		revID(uri2, 1),
 	)
 
 	// create revision 3, and obsolete revision 2.
-	createNewRevision(c, uri1)
-	createNewRevision(c, uri2)
+	createNewRevision(c, st, uri1)
+	createNewRevision(c, st, uri2)
 
 	wCSingleUnit.AssertChange(
 		revID(uri2, 2),
 	)
 
 	wCSingleUnit.AssertNoChange()
 }
+
+func (s *watcherSuite) TestWatchConsumedSecretsChanges(c *gc.C) {
+	s.setupUnits(c, "mysql")
+	s.setupUnits(c, "mediawiki")
+
+	ctx := context.Background()
+	svc, st := s.setupServiceAndState(c)
+
+	saveConsumer := func(uri *coresecrets.URI, revision int, consumerID string) {
+		consumer := &coresecrets.SecretConsumerMetadata{
+			CurrentRevision: revision,
+		}
+		err := st.SaveSecretConsumer(ctx, uri, consumerID, consumer)
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	sp := secret.UpsertSecretParams{
+		Data: coresecrets.SecretData{"foo": "bar", "hello": "world"},
+	}
+	uri1 := coresecrets.NewURI()
+	err := st.CreateCharmApplicationSecret(ctx, 1, uri1, "mysql", sp)
+	c.Assert(err, jc.ErrorIsNil)
+
+	uri2 := coresecrets.NewURI()
+	err = st.CreateCharmApplicationSecret(ctx, 1, uri2, "mysql", sp)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// The consumed revision 1 is the initial revision - will be ignored.
+	saveConsumer(uri1, 1, "mediawiki/0")
+	// The consumed revision 1 is the initial revision - will be ignored.
+	saveConsumer(uri2, 1, "mediawiki/0")
+
+	watcher, err := svc.WatchConsumedSecretsChanges(ctx, "mediawiki/0")
+	c.Assert(err, gc.IsNil)
+	c.Assert(watcher, gc.NotNil)
+	defer workertest.CleanKill(c, watcher)
+
+	wC := watchertest.NewStringsWatcherC(c, watcher)
+
+	// Wait for the initial changes.
+	wC.AssertChange([]string(nil)...)
+	wC.AssertNoChange()
+
+	// create revision 2.
+	createNewRevision(c, st, uri1)
+
+	wC.AssertChange(
+		uri1.String(),
+	)
+	wC.AssertNoChange()
+
+	// pretent that the agent restarted and the watcher is re-created.
+	watcher1, err := svc.WatchConsumedSecretsChanges(ctx, "mediawiki/0")
```

Comment:
Can we also have a test that on restart, if the consumed revision is the same as the latest revision,there's no event